src/lib/Renderer.ts

```diff
@@ -41,11 +43,16 @@ const TIMEOUT = 10000;
 const DATA_REGEXP = /^data:/i;
 
 export interface TaskParams {
+  type: 'render' | 'login';
   url: URL;
   userAgent: string;
   headersToForward: {
     [s: string]: string;
   };
+  login?: {
+    username: string;
+    password: string;
+  };
 }
```

Comment:
Here you allow "render" to have a login prop (which has no side effect but incorrect)
```suggestion
export type TaskParams = ({
  type: 'render';
 | {
  type: 'login';
  login?: {
    username: string;
    password: string;
  };
}) & {   url: URL;
  userAgent: string;
  headersToForward: {
    [s: string]: string;
  } };
```

Reply:
Ok, but I really prefer the permissive version 😬

Reply:
Hmm it doesn't work actually: `Property 'login' does not exist on type 'TaskParams'.`
I'll do a union of two separate types

Reply:
I had made a typo but seems to work, 
https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcAqBDAzgawAoZQYC2WcAvHABQDeAUHAssAFxwDkUwAdgCbBR2AbgYBfOAB849RkhRt2AGwgBzAJbdhDOMvXc2MxnACuWAdxKs4WGFA0qRRuGGxYA7tF5sbd7g+2iIqIAlHAAZNJGxlCKbACqAEoAMo4mZlAAgio8MN629qkAFsAY-FBYaBAAYtBuhF7S2owA2lh5vioAuu0FAXCBdHQAxhDcNggAjGyYuAREpBSRclZKqhrsADQ6a-qRpuaWCmDKYJvOrh5QDewAZhAQ7P1b0bFw3MBucIlJVOzswVtiqUBBVqrV6gZRM90lkcgpHgMRmN4DAAEzTbD4QgkMiUGhMeQcLh8ARnXQaAxpA7EFbHCCnLYuLDuTwKO4PJ4mGJsd6fb6-f6AkplUE1KB1K6Q6ECWHcXIcBEiJHjGAAZgxs2xCzxBJWxLKZxePI+X2SAoBcCBIsqYolDRoUKpmWycvh-REdCAA

Reply:
But then how do you cast it in a function?
```
function test(task: TaskParams) {
  console.log(task.login);
}
```

Reply:
ah then Typescript guess the type with the pivot:
https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcAqBDAzgawAoZQYC2WcAvHABQDeAUHAssAFxwDkUwAdgCbBR2AbgYBfOAB849RkhRt2AGwgBzAJbdhDOMvXc2MxnACuWAdxKs4WGFA0qRRuGGxYA7tF5sbd7g+2iIqIAlHAAZNJGxlCKbACqAEoAMo4mZlAAgio8MN629qkAFsAY-FBYaBAAYtBuhF7S2owA2lh5vioAuu0FAXCBdHQAxhDcNggAjGyYuAREpBSRclZKqhrsADQ6a-qRpuaWCmDKYJvOrh5QDewAZhAQ7P1b0bFw3MBucIlJVOzswVtiqUBBVqrV6gZRM90lkcgpHgMRmN4DAAEzTbD4QgkMiUGhMeQcLh8ARnXQaAxpA7EFbHCCnLYuLDuTwKO4PJ4mGJsd6fb6-f6AkplUE1KB1K6Q6ECWHcXIcBEiJHjGAAZgxs2xCzxBJWxLKZxePI+X2SAoBcCBIsqYolDRoUKpmWycvh-REdBuxm4QxgalGCGANioLnmbXQmLmONChjUN2ooZxADplhRyJRVnp-o0nInSEnydw+sBFGYc0Y81gCzs4AB6WtwARQaBwNRkbjQYgYRQBOiiIA